### PR TITLE
feat: connector enumerator interface for batch

### PR DESCRIPTION
**Pull request by tabVersion**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

***PLEASE DO NOT LEAVE THIS EMPTY !!!***

Please explain **IN DETAIL** what the changes are in this PR and why they are needed:

this pr offers a preview of the interface for batch

will implement in connector trait later

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Refer to a related PR or issue link (optional)

@ZENOTME

### CODE REVIEW

File: src/connector/src/source/kafka/enumerator/client.rs [hunk 1]
```diff
@@ -115,6 +115,69 @@ impl SplitEnumerator for KafkaSplitEnumerator {
...
+    ) -> anyhow::Result<Vec<KafkaSplit>> {
+        let topic_partitions = self.fetch_topic_partition().await.map_err(|e| {
+            anyhow!(format!(
+                "failed to fetch metadata from kafka ({}), error: {}",
+                self.broker_address, e
+            ))
+        })?;
+        let mut expect_start_offset = if let Some(ts) = expect_start_timestamp_millis {
+            Some(
+                self.fetch_offset_for_time(topic_partitions.as_ref(), ts)
```
Comment (liurenjie1024) [inline src/connector/src/source/kafka/enumerator/client.rs:131]: This is waiting for implementation?

File: src/connector/src/source/kafka/enumerator/client.rs [hunk 2]
```diff
@@ -115,6 +115,69 @@ impl SplitEnumerator for KafkaSplitEnumerator {
...
+        let mut expect_stop_offset = if let Some(ts) = expect_stop_timestamp_millis {
+            Some(
+                self.fetch_offset_for_time(topic_partitions.as_ref(), ts)
+                    .await?,
+            )
+        } else {
+            None
+        };
+
+        let mut watermarks = {
```
Comment (liurenjie1024) [inline src/connector/src/source/kafka/enumerator/client.rs:146]: Why we need watermarks here?

File: src/connector/src/source/kafka/enumerator/client.rs [hunk 3]
```diff
@@ -115,6 +115,69 @@ impl SplitEnumerator for KafkaSplitEnumerator {
...
+                i64::min(stop, high)
+            };
+
+            if start_offset > stop_offset {
+                return Err(anyhow!(format!("topic {} partition {}: requested start offset {} is greater than stop offset {}",self.topic, partition, start_offset, stop_offset)));
+            }
+            Ok(KafkaSplit {
+                topic: self.topic.clone(),
+                partition: *partition,
+                start_offset: Some(start_offset),
```
Comment (liurenjie1024) [inline src/connector/src/source/kafka/enumerator/client.rs:175]: If we always return some offset, why not remove the `Option`?

File: src/connector/src/source/kafka/enumerator/client.rs (hunk 2 above)
Comment (liurenjie1024) [reply src/connector/src/source/kafka/enumerator/client.rs:146]: Please add some comments here, the name `watermark` is quite confusing.

**Review by liurenjie1024 — COMMENTED** [review]
Generally LGTM

File: src/connector/src/source/kafka/enumerator/client.rs (hunk 1 above)
Comment (tabVersion) [reply src/connector/src/source/kafka/enumerator/client.rs:131]: we will only support kafka for now. Other connectors are pending

File: src/connector/src/source/kafka/enumerator/client.rs (hunk 2 above)
Comment (tabVersion) [reply src/connector/src/source/kafka/enumerator/client.rs:146]: watermark means the earliest and the latest message index in each partition. Will describe details here

File: src/connector/src/source/kafka/enumerator/client.rs (hunk 3 above)
Comment (tabVersion) [reply src/connector/src/source/kafka/enumerator/client.rs:175]: To align with other connectors. Still it is a good suggestion, I will take a deeper look

**Review by ZENOTME — APPROVED** [review]

**Pull request merged**